### PR TITLE
Add tool output truncation middleware

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Add configurable truncation middleware for shell, file, and search tool outputs.
- Store full over-limit output in ArtifactStore and return compact text with artifact reference.
- Wire tests into build and cover agent-loop byte telemetry with artifact-backed output.

## Test plan
- [x] cd zig && zig build test

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: REQUEST_CHANGES**

Overall the implementation is clean and well-structured. The middleware signature matches `ToolOutputMiddlewareFn` exactly, the agent_loop integration is correct, and all acceptance criteria are covered by tests. Found one P1 correctness issue and a few P2/P3 items.

---

### P1 — Memory leak in `appendArtifactReference` error path

**File:** `zig/src/tools/truncation.zig:121-130`

```zig
fn appendArtifactReference(...) ![]ai_types.ArtifactReference {
    const out = try allocator.alloc(ai_types.ArtifactReference, existing.len + 1);
    errdefer allocator.free(out);

    for (existing, 0..) |artifact, i| {
        out[i] = try cloneArtifactReference(allocator, artifact);
    }
```

If `cloneArtifactReference` fails at iteration `i > 0`, the errdefer frees the outer array but **does not** deinit the already-cloned entries (`out[0..i]`). Each cloned `ArtifactReference` contains duped strings (`artifact_id`, `uri`, `mime_type`, `sha256`, `description`) that all leak. 

**Fix:** Track progress and deinit partial clones:
```zig
var cloned_count: usize = 0;
errdefer {
    for (out[0..cloned_count]) |*a| a.deinit(allocator);
    allocator.free(out);
}
for (existing, 0..) |artifact, i| {
    out[i] = try cloneArtifactReference(allocator, artifact);
    cloned_count += 1;
}
```

In practice, `result.artifacts` is usually empty when the middleware runs, making this path rare. But it's a latent leak that `std.testing.allocator` would catch if the path were exercised.

---

### P2 — `collectToolOutput` mixes text + `details_json` in truncation boundary

**File:** `zig/src/tools/truncation.zig:93-114`

`collectToolOutput` concatenates all text content parts **and then appends `details_json`** into a single flat buffer. The `limit` check and truncation slice operate on this combined buffer. This means:

1. The truncation boundary may split mid-JSON, producing corrupted details in the truncated output.
2. The stored artifact content is a concatenation of text + raw JSON, which is not a useful artifact to retrieve.
3. If `details_json` is large, text content gets truncated earlier than expected.

**Suggestion:** Either exclude `details_json` from the collected output (the middleware already replaces it with an empty string at line 80), or at minimum document this behavior. The truncation should arguably operate on text content only.

---

### P2 — No test for `fallback = null` / unrecognized tool type pass-through

When `Limits.fallback` is `null` (the default) and the tool name doesn't match any pattern, `forTool` returns `null` → `apply` returns early → no truncation. This is correct behavior but has **no test coverage**. Add a test with an unrecognized tool name and `fallback = null` to confirm zero-overhead pass-through.

---

### P3 — `classifyTool` substring matching is fragile

**File:** `zig/src/tools/truncation.zig:86-91`

`std.mem.indexOf(u8, tool_name, "file")` matches any tool name *containing* "file" as a substring. Examples of false positives:
- `"profile_read"` → matches `.file` (contains "file" in "profile")
- `"seashell_analyzer"` → matches `.shell` (contains "shell" in "seashell")

Works for current tool names but will bite if naming conventions drift. Consider matching against known tool name prefixes or using exact comparison against a set.

---

### Test Results

`zig build test` — all pass (local). CI pending.

File: zig/src/tools/truncation.zig [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+fn truncateWithMarker(allocator: std.mem.Allocator, raw_output: []const u8, limit: usize, artifact_id: []const u8) ![]u8 {
+    const head_len = @min(limit, raw_output.len);
+    return std.fmt.allocPrint(allocator, "{s}" ++ marker_prefix ++ "{s}" ++ marker_suffix, .{ raw_output[0..head_len], artifact_id[0..@min(artifact_id.len, 12)] });
+}
+
+fn appendArtifactReference(allocator: std.mem.Allocator, existing: []const ai_types.ArtifactReference, new_reference: ai_types.ArtifactReference) ![]ai_types.ArtifactReference {
+    const out = try allocator.alloc(ai_types.ArtifactReference, existing.len + 1);
+    errdefer allocator.free(out);
```
Comment (lsm) [inline zig/src/tools/truncation.zig]: **P1 — Memory leak on error path.** If `cloneArtifactReference` fails at iteration `i > 0`, the errdefer frees the outer array but does not deinit already-cloned entries (`out[0..i]`). Each cloned `ArtifactReference` contains duped strings that leak.

Fix: track `cloned_count` and deinit partial clones in the errdefer:
```zig
var cloned_count: usize = 0;
errdefer {
    for (out[0..cloned_count]) |*a| a.deinit(allocator);
    allocator.free(out);
}
for (existing, 0..) |artifact, i| {
    out[i] = try cloneArtifactReference(allocator, artifact);
    cloned_count += 1;
}
```

File: zig/src/tools/truncation.zig [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    return .other;
+}
+
+fn collectToolOutput(allocator: std.mem.Allocator, result: AgentToolResult) ![]u8 {
+    var total: usize = 0;
+    for (result.content.slice()) |part| switch (part) {
+        .text => |text| total += text.text.len,
+        .image => {},
+    };
+    if (result.getDetailsJson()) |details| total += details.len;
```
Comment (lsm) [inline zig/src/tools/truncation.zig]: **P2 - collectToolOutput appends details_json into the combined buffer, making the truncation boundary unpredictable.** The stored artifact content becomes text + raw_json, which is not useful for retrieval. Consider excluding details_json from collection (the middleware already replaces it with empty at line 80), or document this behavior.

File: zig/src/tools/truncation.zig [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        result.* = .{
+            .content = ai_types.OwnedSlice(ai_types.UserContentPart).initOwned(new_content),
+            .details_json = ai_types.OwnedSlice(u8).initBorrowed(""),
+            .artifacts = ai_types.OwnedSlice(ai_types.ArtifactReference).initOwned(artifacts),
+        };
+    }
+};
+
+fn classifyTool(tool_name: []const u8) ToolKind {
+    if (std.mem.indexOf(u8, tool_name, "shell") != null) return .shell;
```
Comment (lsm) [inline zig/src/tools/truncation.zig]: **P3 - classifyTool uses substring matching via std.mem.indexOf.** This means 'profile_read' matches .file (contains 'file' in 'profile'), and 'seashell_analyzer' matches .shell. Works for current tool names but fragile against naming drift. Consider matching against known tool name prefixes or exact comparison.

File: zig/src/tools/truncation.zig [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    };
+    if (result.getDetailsJson()) |details| {
+        @memcpy(out[offset .. offset + details.len], details);
+    }
+    return out;
+}
+
+fn truncateWithMarker(allocator: std.mem.Allocator, raw_output: []const u8, limit: usize, artifact_id: []const u8) ![]u8 {
+    const head_len = @min(limit, raw_output.len);
+    return std.fmt.allocPrint(allocator, "{s}" ++ marker_prefix ++ "{s}" ++ marker_suffix, .{ raw_output[0..head_len], artifact_id[0..@min(artifact_id.len, 12)] });
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: marker now counts against configured byte limit, so returned text length stays <= limit; test now asserts this bound.

File: zig/src/tools/truncation.zig [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        errdefer reference.deinit(allocator);
+
+        const truncated = try truncateWithMarker(allocator, raw_output, limit, reference.artifact_id);
+        errdefer allocator.free(truncated);
+
+        const new_content = try allocator.alloc(ai_types.UserContentPart, 1);
+        errdefer allocator.free(new_content);
+        new_content[0] = .{ .text = .{ .text = truncated } };
+
+        const artifacts = try appendArtifactReference(allocator, result.artifacts.slice(), reference);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: store.write reference is cloned into middleware allocator before insertion, and store-owned reference is deinitialized with store allocator.

File: zig/src/tools/truncation.zig (hunk 1 above)
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: appendArtifactReference now tracks cloned_count and deinitializes partial clones on error before freeing outer slice.

File: zig/src/tools/truncation.zig (hunk 2 above)
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: collectToolOutput now includes only text content, excluding details_json from stored artifact/truncation boundary.

File: zig/src/tools/truncation.zig (hunk 3 above)
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: classifyTool now matches exact tool names or token prefixes followed by _/-; profile_read no longer matches file and added fallback=null no-op test.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: APPROVE**

All prior findings addressed in this round:

| Finding | Status | Fix |
|---------|--------|-----|
| P1: Memory leak in `appendArtifactReference` errdefer | ✅ Fixed | `cloned_count` tracking + partial deinit (lines 132-136) |
| P2: `collectToolOutput` mixed details_json | ✅ Fixed | Removed details_json from collection (lines 104-121) |
| P2: No test for fallback=null pass-through | ✅ Fixed | New test "ignores unknown tools without fallback" (line 268) |
| P3: `classifyTool` substring matching | ✅ Fixed | New `matchesToolName` with prefix+separator check (lines 96-102) |

**Additional improvements noted:**
- Store reference ownership properly separated: `store_reference` deinited via `defer`, cloned `reference` goes into result
- `truncateWithMarker` now accounts for marker length, ensuring total output ≤ limit
- Test assertions strengthened (size bounds, exact head length computation)

No new findings. Tests pass locally. Clean approval.

File: zig/src/tools/truncation.zig [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            allocator.free(artifacts);
+        }
+
+        result.content.deinit(allocator);
+        result.details_json.deinit(allocator);
+        result.artifacts.deinit(allocator);
+
+        result.* = .{
+            .content = ai_types.OwnedSlice(ai_types.UserContentPart).initOwned(new_content),
+            .details_json = ai_types.OwnedSlice(u8).initBorrowed(""),
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: details_json is preserved during truncation so shell ok/exit_code metadata remains available to model/event consumers.

File: zig/src/tools/truncation.zig [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    const separator = tool_name[token.len];
+    return separator == '_' or separator == '-';
+}
+
+fn collectToolOutput(allocator: std.mem.Allocator, result: AgentToolResult) ![]u8 {
+    var total: usize = 0;
+    for (result.content.slice()) |part| switch (part) {
+        .text => |text| total += text.text.len,
+        .image => {},
+    };
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: truncation decision now uses text bytes + details_json bytes, and text budget is reduced by details_json length so returned total stays within configured limit.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: APPROVE**

All review feedback addressed across three rounds. No remaining findings.

### Round 3 Changes Verified

- `details_json` now preserved in result (not discarded) and counted toward byte limit via `raw_output.len + details.len <= limit`
- Ownership transfer via `@constCast` for already-owned `details_json` — no error path between transfer and reassignment, safe
- Text budget correctly reduced by `details.len`: `limit -| details.len` with saturating subtraction
- New test "preserves details and counts them toward limit" validates the behavior
- `truncateWithMarker` handles edge cases (limit=0 → empty, limit ≤ marker → truncated marker)
- Integration test handles empty-text case with `or text.len == 0` fallback

### Summary of All Rounds

| Finding | Round | Status |
|---------|-------|--------|
| P1: Memory leak in `appendArtifactReference` errdefer | R1 | ✅ Fixed |
| P2: `collectToolOutput` mixed details_json into artifact | R1 | ✅ Fixed |
| P2: No fallback=null test | R1 | ✅ Fixed |
| P3: `classifyTool` substring matching | R1 | ✅ Fixed |
| details_json preservation + budget accounting | R2→R3 | ✅ Implemented |

All tests pass locally. Clean approval.

File: zig/src/tools/truncation.zig [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+            .content = raw_output,
+            .mime_type = mime_text_plain,
+            .description = "raw tool output",
+        });
+        defer store_reference.deinit(self.store.allocator);
+
+        var reference = try cloneArtifactReference(allocator, store_reference);
+        errdefer reference.deinit(allocator);
+
+        const truncated = try truncateWithMarker(allocator, raw_output, limit -| details.len, reference.artifact_id);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/truncation.zig]: Fixed: details_json now truncates too when it exceeds the configured limit. Returned text budget is computed after details truncation, and a new test covers details-only over-limit payloads.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: APPROVE**

All review feedback addressed across four rounds. No remaining findings.

### Round 4 Changes Verified

- **`truncateDetailsJson`** — new function truncates details_json when it alone exceeds limit, reusing same marker format
- **Simplified ownership** — no more `@constCast` ownership transfer; details_json is always freshly allocated; `result.details_json.deinit(allocator)` properly called
- **Budget split** — details truncated first, then `text_budget = limit -| returned_details.len`
- **New test** "truncates details that exceed limit" — validates large details get truncated and total payload stays capped
- **Integration test** works correctly with fallback=0 (both text and details truncated to empty)

### Cumulative Summary

| Finding | Round | Status |
|---------|-------|--------|
| P1: Memory leak in `appendArtifactReference` errdefer | R1 | ✅ Fixed |
| P2: `collectToolOutput` mixed details_json | R1 | ✅ Fixed |
| P2: No fallback=null test | R1 | ✅ Fixed |
| P3: `classifyTool` substring matching | R1 | ✅ Fixed |
| details_json preservation + budget | R2→R3 | ✅ Implemented |
| details_json truncation when exceeding limit | R3→R4 | ✅ Implemented |

All tests pass. All review threads resolved. Clean approval.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: REQUEST_CHANGES**

After 4 rounds of fixes, the implementation is technically correct. However, a significant architectural issue emerged during review: **this PR duplicates functionality already shipped in PR #97**.

### P1 — Duplicate artifact-backing paths

PR #97 already implements tool-output artifact storage in `common.makeTextResultWithArtifact()` (`zig/src/tools/common.zig:299-340`):
- Auto-stores output >4KB via `storeArtifact()`
- Returns head+tail summary + artifact path + byte telemetry
- Populates `ArtifactReference` on result
- `artifact_retrieve` tool for retrieval

PR #101 adds `TruncationMiddleware` that does the same thing at a different layer:
- Stores output via `ArtifactStore`
- Returns first-N-bytes + marker

Both can fire on the same output → **double-storage**.

### Consolidation required

Keep tool-side (PR #97) as primary — its head+tail summary is better UX. Enhance it with:
1. Per-tool-type configurable limits from this PR's `Limits` struct → port into `common.zig`
2. `ArtifactStore` backend (from PR #100) → replace file-based `storeArtifact` when available
3. Keep `ToolOutputMiddlewareFn` hook as extension point
4. Delete `TruncationMiddleware` or reduce to guard that skips when result already has artifacts

Detailed plan sent to coding agent. Full context in message.

**Comment by lsm** [comment]
Addressed consolidation review:

- Removed duplicate `TruncationMiddleware` module and build/test wiring.
- Moved per-tool output limits into `tools/common.zig` with hardened tool classification.
- Extended `makeTextResultWithArtifact()` to use `ArtifactStore` when provided and keep existing file artifact fallback otherwise.
- Kept tool-side head/tail summaries as the primary large-output UX.
- Left `ToolOutputMiddlewareFn` hook intact and updated its integration test to cover generic middleware byte telemetry without artifact duplication.

Validation: `cd zig && zig build test` passes locally.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: APPROVE**

Consolidation done well. `truncation.zig` deleted, per-tool limits and optional `ArtifactStore` backend moved into `common.zig`, head+tail summary preserved as primary UX, middleware hook preserved as extension point. Double-storage risk eliminated. All tests pass.

### What was done

- `ToolOutputLimits` struct in `common.zig` with configurable per-tool-type limits (shell: 10KB, file: 20KB, search: 15KB, fallback: 4KB)
- `classifyTool`/`matchesToolName` ported with prefix+separator matching
- `TextResultOptions.limits` + optional `store: ?*ArtifactStore` fields
- `makeTextResultWithArtifact` uses `limits.forTool()` for threshold; `ArtifactStore` backend when `store` provided, file-based fallback otherwise
- `cloneArtifactReference` for bridging store-allocated references into caller-owned results
- `truncation.zig` deleted entirely
- Integration test simplified to test middleware hook without artifact dependency
- 3 new unit tests: per-tool classification, per-tool limits, ArtifactStore backend

### Minor notes (non-blocking)

- **Shell overrides limit to 4KB** (`shell.zig:102` passes `.limits = .{ .shell = common.default_fallback_limit }`), preserving old behavior instead of using the new 10KB default. Acceptable as a conservative choice but contradicts task spec.
- **File tools don't use `makeTextResultWithArtifact`** (`file.zig` uses `makeTextResultOwned` instead) — no artifact backing for file output. Pre-existing gap from PR #97, not introduced here.
- **ArtifactStore backend is opt-in** — no tool passes `store` yet. Backend is ready for future wiring.

Clean approval. Consolidation achieved.

**Comment by lsm** [comment]
Follow-up fixes applied:

- Shell now uses the new 10KB default limit instead of overriding to the legacy 4KB fallback.
- `file_read` now routes hashed read output through `makeTextResultWithArtifact()`, so large file reads produce artifact references.
- Added file-read large-output artifact coverage and adjusted shell large-output test size for the 10KB threshold.

Validation: `cd zig && zig build test` passes locally.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** Claude Sonnet 4 | **Client:** NeoKai | **Provider:** Anthropic

**Verdict: APPROVE**

Both follow-up items addressed. Clean approval.

### Fixes verified

- **Shell limit** — `shell.zig:102` now uses default `ToolOutputLimits` → shell gets 10KB per spec. No more 4KB override.
- **`file_read` routed through `makeTextResultWithArtifact`** — `file.zig:50` now uses shared artifact helper → large file reads get artifact refs and head/tail summary with 20KB limit.

### Final state

Single artifact-backing path via `common.makeTextResultWithArtifact`. Per-tool-type limits (shell: 10KB, file: 20KB, search: 15KB). Optional `ArtifactStore` backend. Head+tail summary. Middleware hook preserved. No double-storage. All tests pass.
